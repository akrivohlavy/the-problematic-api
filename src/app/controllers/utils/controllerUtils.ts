import { compose, Middleware } from 'compose-middleware';
import { NextFunction, Request, Response } from 'express';
import { NO_CONTENT, OK } from 'http-status-codes';
import { mapValues, omit, values } from 'lodash';
import httpContext from './httpContext';

type Handler = (req: Request, res: Response, next: NextFunction) => any;
type SimpleHandler = (req: Request, res: Response) => any;
type WriteResponse = (req: Request, res: Response, data?: any) => any;
const writeResponse: WriteResponse = (_req, res, data) => res.json(data);

export const respond = (
    controllerHandler: SimpleHandler,
    statusCode = OK,
    respondFn: WriteResponse = writeResponse
): Handler => async (req, res, next) => {
    try {
        const result = await controllerHandler(req, res);
        res.status(statusCode);
        return respondFn(req, res, result);
    } catch (error) {
        next(error);
    }
};

export const respondWithEmpty = (_req: Request, res: Response, data?: any) => {
    if (Array.isArray(data) && data.length === 0) {
        res.status(NO_CONTENT);
        return res.end();
    }
    return res.json(data);
};

export const omitOrder = (o: any) => omit(o, ['order']);
export const omitPagination = (o: any) => omit(o, ['limit', 'offset']);

export const pipeMiddleware = (...middlewares: Array<Middleware<Request, Response>>) => compose(middlewares);

export const bindContext = (req: Request, res: Response, next: NextFunction) => {
    req.context = httpContext({ req, res });
    next();
};

export const meTranslate = (req: Request, res: Response, next: NextFunction) => {
    const context = req.context || httpContext({ req, res });

    if (!context.user) {
        return next();
    }

    const translate = (object: any): any =>
        mapValues(object, (value: any) => {
            if (Array.isArray(value)) {
                return values(translate(value));
            }
            return value === 'me' ? context.user.id : value;
        });

    req.params = translate(req.params);
    req.query = translate(req.query);

    next();
};
