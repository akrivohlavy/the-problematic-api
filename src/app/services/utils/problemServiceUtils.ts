import Problem, { Solution } from '../../models/Problem';

export const problemToSchema = ({ id, type, query, createdBy }: Problem) => ({
    id,
    type,
    query,
    createdBy,
});

export const safeCompareAnswers = (userAnswer: Solution, correctAnswer: Solution): boolean => {
    if (typeof correctAnswer === 'number') {
        return correctAnswer === +userAnswer;
    }
    if (typeof correctAnswer === 'string') {
        const compareOptions = {
            sensitivity: 'base',
            ignorePunctuation: true,
        };

        return (correctAnswer as string).localeCompare(userAnswer as string, undefined, compareOptions) === 0;
    }
    throw new TypeError(`Cannot safely compare unexpected type: ${typeof correctAnswer}`);
};

export const filterAnswered = (keepAnswered: boolean) => (problem: Problem): boolean => {
    const userAnswersCount = problem.Answers.length || 0;
    return keepAnswered ? userAnswersCount > 0 : userAnswersCount === 0;
};
