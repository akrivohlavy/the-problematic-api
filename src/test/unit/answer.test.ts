import { safeCompareAnswers } from '../../app/services/utils/problemServiceUtils';

describe('Problem Service (Unit)', () => {
    describe('safe answer comparison', () => {
        it('casts user answer to number', () => {
            expect(safeCompareAnswers('42', 42)).toBeTrue();
            expect(safeCompareAnswers('42', 1337)).toBeFalse();
        });

        it('handles negative numbers', () => {
            expect(safeCompareAnswers('-9000', -9000)).toBeTrue();
        });

        it('compares exactly matching strings', () => {
            expect(safeCompareAnswers('No problem.', 'No problem.')).toBeTrue();
            expect(safeCompareAnswers('This might be a problem.', 'No problem.')).toBeFalse();
        });

        it('compares strings case insensitively', () => {
            expect(safeCompareAnswers('no problem.', 'NO PROBLEM.')).toBeTrue();
        });

        it('ignores whitespaces and punctuation', () => {
            expect(safeCompareAnswers(' no \tproblem! ', 'No problem.')).toBeTrue();
        });
    });
});
