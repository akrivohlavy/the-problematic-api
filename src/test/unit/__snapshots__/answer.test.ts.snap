// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Problem Service (Unit) filtering list of Problems based on user answers filters answered question 1`] = `
Array [
  Object {
    "Answers": Array [
      Object {
        "ProblemId": 2,
        "createdAt": "2019-09-28T01:43:12.399Z",
        "id": 1,
        "updatedAt": "2019-09-28T01:43:12.401Z",
        "username": "meksi",
      },
    ],
    "createdAt": "2019-09-28T01:42:29.350Z",
    "createdBy": "tonda",
    "id": 2,
    "query": "2 + 2",
    "type": "expression",
    "updatedAt": "2019-09-28T01:42:29.350Z",
  },
]
`;

exports[`Problem Service (Unit) filtering list of Problems based on user answers filters unanswered question 1`] = `
Array [
  Object {
    "Answers": Array [],
    "createdAt": "2019-09-28T01:42:28.024Z",
    "createdBy": "tonda",
    "id": 1,
    "query": "U WOT M8?",
    "type": "riddle",
    "updatedAt": "2019-09-28T01:42:28.024Z",
  },
  Object {
    "Answers": Array [],
    "createdAt": "2019-09-28T01:43:05.504Z",
    "createdBy": "tonda",
    "id": 3,
    "query": "What time is it?",
    "type": "riddle",
    "updatedAt": "2019-09-28T01:43:05.504Z",
  },
]
`;
